### PR TITLE
Add Blobstore Library

### PR DESCRIPTION
**Pull request by difanyang**
I used a new object called Input to store both the comment and the image url. When running the dev server, I keep getting the same error message "Error 405 HTTP method GET is not supported by this URL". I have properly overridden all my doGet() and doPost() functions. What could be the problem?

### CODE REVIEW

File: portfolio/src/main/java/com/google/sps/servlets/BlobstoreUploadUrlServlet.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+ * When the fetch() function requests the /blobstore-upload-url URL, the content of the response is
+ * the URL that allows a user to upload a file to Blobstore. If this sounds confusing, try running a
+ * dev server and navigating to /blobstore-upload-url to see the Blobstore URL.
+ */
+@WebServlet("/blobstore-upload-url")
+public class BlobstoreUploadUrlServlet extends HttpServlet {
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    BlobstoreService blobstoreService = BlobstoreServiceFactory.getBlobstoreService();
```
Comment (zixiyan) [inline portfolio/src/main/java/com/google/sps/servlets/BlobstoreUploadUrlServlet.java:21]: consider making it private static final?

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 2]
```diff
@@ -34,31 +48,43 @@
...
+    Query query = new Query("Input");
     PreparedQuery results = DATASTORE.prepare(query);
-    ArrayList<String> stringList = new ArrayList<String>();
+    ArrayList<Input> inputList = new ArrayList<Input>();
     int numComment = Integer.parseInt(getParameterWithDefault(request, "numComment", "0"));
 
     for (final Entity entity : results.asIterable(FetchOptions.Builder.withLimit(numComment))) {
-      stringList.add((String) entity.getProperty("content"));
+      inputList.add(new Input((String)entity.getProperty("comment"),
+      (String)entity.getProperty("imageUrl")));
```
Comment (zixiyan) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:62]: please format. try to align these parameters.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 3]
```diff
@@ -73,4 +99,39 @@ private String getParameterWithDefault(HttpServletRequest request, String name,
...
+
+    // User submitted form without selecting a file, so we can't get a URL. (live server)
+    BlobInfo blobInfo = new BlobInfoFactory().loadBlobInfo(blobKey);
+    if (blobInfo.getSize() == 0) {
+      BLOBSTORE.delete(blobKey);
+      return null;
+    }
+
+    // Use ImagesService to get a URL that points to the uploaded file.
+    ImagesService imagesService = ImagesServiceFactory.getImagesService();
```
Comment (zixiyan) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:125]: is it possible to make it private static final?

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 4]
```diff
@@ -34,31 +48,44 @@
 public class DataServlet extends HttpServlet {
   private static final Gson GSON = new Gson();
   private static final DatastoreService DATASTORE = DatastoreServiceFactory.getDatastoreService();
+  private static final BlobstoreService BLOBSTORE = BlobstoreServiceFactory.getBlobstoreService();
+  private static final ImagesService IMAGESERVICE = ImagesServiceFactory.getImagesService();
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    Query query = new Query("Comment");
+    Query query = new Query("Input");
```
Comment (zixiyan) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:56]: I would suggest make constants like ```Input``` to private static final. In such a way, we could catch some simple typo.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 5]
```diff
@@ -73,4 +100,38 @@ private String getParameterWithDefault(HttpServletRequest request, String name,
     }
     return value;
   }
+
+   /** Returns a URL that points to the uploaded file, or null if the user didn't upload a file. */
+  private String getUploadedFileUrl(HttpServletRequest request, String formInputElementName) {
```
Comment (zixiyan) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:105]: Seems that you don't need ```formInputElementName```?

**Comment by zixiyan** [comment]
curious, what's the bug?

**Comment by difanyang** [comment]
> curious, what's the bug?

The webpage would give Error 405 “http method get is not supported by this url” after the form was submitted. At first I thought it was the Blobstore failing to create the right upload url, but the actual fix was changing the redirecting url in doPost() from "index.html" to "/index.html".

**Pull request merged**